src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -135,9 +135,10 @@ public GeneratorDriverRunResult GetRunResult()
                                 => new GeneratorRunResult(generator,
                                                           diagnostics: generatorState.Diagnostics,
                                                           exception: generatorState.Exception,
-                                                          generatedSources: generatorState.SourceTexts.ZipAsArray(
-                                                                                generatorState.Trees,
-                                                                                (sourceText, tree) => new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName))));
+                                                          generatedSources: generatorState.PostInitTrees
+                                                                              .Union(generatorState.GeneratedTrees)
+                                                                              .Select(t => new GeneratedSourceResult(t.Tree, t.Text, t.HintName))
+                                                                              .ToImmutableArray()));
```

Comment:
Consider using an `ArrayBuilder` and explicitly adding entries to reduce the allocations.